### ISSUE TITLE
kube-dns-v2: images not found

### ISSUE THREAD

**Issue by bkeroackdsc** [GoogleCloudPlatform/kubernetes#9324]
I'm running k8s 0.18.1.1 on CoreOS (alpha) in AWS.

I'm trying to get the DNS pod working per the [documentation](https://github.com/GoogleCloudPlatform/kubernetes/tree/master/cluster/addons/dns).

The container images in the [example controller](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/addons/dns/skydns-rc.yaml.in) do not appear to exist:

```
dockerd[1091]: time="2015-06-05T17:33:54Z" level=error msg="HTTP Error: statusCode=404 No such image: gcr.io/google_containers/skydns:2015-03-11-001"

dockerd[1091]: time="2015-06-05T17:33:54Z" level=error msg="HTTP Error: statusCode=404 No such image: gcr.io/google_containers/kube2sky:1.8"
```

Changing the tags to "latest" does not work either:

```
dockerd[1091]: time="2015-06-05T17:33:54Z" level=error msg="HTTP Error: statusCode=404 No such image: gcr.io/google_containers/skydns:latest"

dockerd[1091]: time="2015-06-05T17:33:54Z" level=error msg="HTTP Error: statusCode=404 No such image: gcr.io/google_containers/kube2sky:latest"
```

It seems that gcr.io only supports a subset of the v1 Registry API, so it's difficult to determine a valid tag or if those repositories even exist:

```bash
 http GET https://gcr.io/v1/repositories/google_containers/kube2sky/tags
HTTP/1.1 401 Unauthorized
Alternate-Protocol: 443:quic,p=1
Content-Length: 0
Content-Type: text/plain; charset=UTF-8
Date: Fri, 05 Jun 2015 17:31:21 GMT
Server: Docker Registry
WWW-Authenticate: Token realm=""
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block

$ http GET https://gcr.io/v1/repositories/google_containers/kube2sky/tags/latest
HTTP/1.1 405 Method Not Allowed
Alternate-Protocol: 443:quic,p=1
Content-Length: 0
Content-Type: text/plain; charset=UTF-8
Date: Fri, 05 Jun 2015 17:27:35 GMT
Server: Docker Registry
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block

$ http GET https://gcr.io/v1/search
HTTP/1.1 404 Not Found
Alternate-Protocol: 443:quic,p=1
Content-Length: 1434
Content-Type: text/html; charset=UTF-8
Date: Fri, 05 Jun 2015 17:28:26 GMT
Server: Docker Registry
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
```

Where should I be pulling these images from?

**Comment by dchen1107** [comment]
cc/ @thockin @brendandburns

**Comment by dchen1107** [comment]
cc @ArtfulCoder

**Comment by thockin** [comment]
We explicitly do not support :latest, but kube2sky:1.8 works for me.

```
$ docker pull gcr.io/google_containers/kube2sky:1.8
Pulling repository gcr.io/google_containers/kube2sky
ce5eb3b04bea: Download complete
511136ea3c5a: Download complete
df7546f9f060: Download complete
ea13149945cb: Download complete
4986bf8c1536: Download complete
041462d0f6a3: Download complete
b1a016807675: Download complete
ee06e4e3ffe6: Download complete
Status: Image is up to date for gcr.io/google_containers/kube2sky:1.8

$ docker pull gcr.io/google_containers/skydns:2015-03-11-001
Pulling repository gcr.io/google_containers/skydns
791ddf327076: Download complete
511136ea3c5a: Download complete
df7546f9f060: Download complete
ea13149945cb: Download complete
4986bf8c1536: Download complete
26151bfe665f: Download complete
59a46ecc9803: Download complete
Status: Downloaded newer image for
gcr.io/google_containers/skydns:2015-03-11-001
```

@brendandburns any chance your recent changes to this bucket might have
caused problems?

**Comment by ArtfulCoder** [comment]
@bkeroackdsc I was able to pull the docker image as well from a guest network where I was not logged into any google accounts (gcloud auth revoke)

I did get the 401 and 405 http response codes as well, but inspite of that the docker pull command worked. So the http response codes might be a red herring

**Comment by bkeroackdsc** [comment]
I thought perhaps the local docker registry mirror might have been interfering somehow, so I destroyed and spun up my cluster without it.

I'm still getting these errors in the journal:

```
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="GET /images/gcr.io/google_containers/kube2sky:1.8/json"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="+job image_inspect(gcr.io/google_containers/kube2sky:1.8)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="GET /containers/d2cc9ef56fc3cac4c38784da9cfa1eed0f296f1566eff50e9b24d03f80ffe88d/json"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="+job container_inspect(d2cc9ef56fc3cac4c38784da9cfa1eed0f296f1566eff50e9b24d03f80ffe88d)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="-job container_inspect(d2cc9ef56fc3cac4c38784da9cfa1eed0f296f1566eff50e9b24d03f80ffe88d) = OK (0)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: No such image: gcr.io/google_containers/kube2sky:1.8
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="-job image_inspect(gcr.io/google_containers/kube2sky:1.8) = ERR (1)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=error msg="Handler for GET /images/{name:.*}/json returned error: No such image: gcr.io/google_containers/kube2sky:1.8"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=error msg="HTTP Error: statusCode=404 No such image: gcr.io/google_containers/kube2sky:1.8"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="GET /containers/0a01cc5b7676b2083bb4c3a8e610fc7ae9faa763a4fd0b2a45f3b1ac7fcc7f21/json"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="+job container_inspect(0a01cc5b7676b2083bb4c3a8e610fc7ae9faa763a4fd0b2a45f3b1ac7fcc7f21)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="-job container_inspect(0a01cc5b7676b2083bb4c3a8e610fc7ae9faa763a4fd0b2a45f3b1ac7fcc7f21) = OK (0)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="GET /images/gcr.io/google_containers/skydns:2015-03-11-001/json"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="+job image_inspect(gcr.io/google_containers/skydns:2015-03-11-001)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: No such image: gcr.io/google_containers/skydns:2015-03-11-001
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=info msg="-job image_inspect(gcr.io/google_containers/skydns:2015-03-11-001) = ERR (1)"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=error msg="Handler for GET /images/{name:.*}/json returned error: No such image: gcr.io/google_containers/skydns:2015-03-11-001"
Jun 05 20:32:00 ip-10-10-28-33.domain.com dockerd[685]: time="2015-06-05T20:32:00Z" level=error msg="HTTP Error: statusCode=404 No such image: gcr.io/google_containers/skydns:2015-03-11-001"
```

Weirdlly, manually pulling the images works:
```bash
$ docker pull gcr.io/google_containers/skydns:2015-03-11-001
Pulling repository gcr.io/google_containers/skydns
791ddf327076: Download complete
511136ea3c5a: Download complete
df7546f9f060: Download complete
ea13149945cb: Download complete
4986bf8c1536: Download complete
26151bfe665f: Download complete
59a46ecc9803: Download complete
Status: Downloaded newer image for gcr.io/google_containers/skydns:2015-03-11-001
```

Any ideas?

**Comment by thockin** [comment]
We did see something like this a LONG time ago.  @dchen1107 do you recall
what this was?

**Comment by dchen1107** [comment]
@thockin, we saw an issue with similar symptom before. The root cause is the old implementation of image garbage collection removed untagged images periodically, and unfortunately all intermediate images are untagged during the pulling. In this case, there are a race between pulling and rmi. We fixed image garbage collection long time ago, and haven't touched it for a while. 

@bkeroackdsc Could you please copy & paste related kubelet log too? Thanks!

**Comment by bkeroackdsc** [comment]
I spent some more time on this and I think it's working now. I think the root issue was the secret volume that the RC template is using. After attempting to create the "token-system-dns" secret with the proper format (apparently it is interpreted as YAML and used for k8s client configuration), I gave up and removed the secret altogether and specified the master URL directly. It failed at first (``Failed to list *api.Service: Get https://x.x.x.x:443/api/v1beta3/services: x509: failed to load system roots and no roots provided``) but I switched to the insecure port and now it seems to be running fine.

**Comment by maklemenz** [comment]
@bkeroackdsc can you please share the configuration you came up with?

**Comment by bkeroackdsc** [comment]
@maklemenz My cluster is running on CoreOS in AWS, more-or-less based on [the docs](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/getting-started-guides/coreos/coreos_multinode_cluster.md). I noticed that after the cluster came online I had a "kube-dns" service but no corresponding controllers or pods.

I began using the [example controller](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/addons/dns/skydns-rc.yaml.in). After working through various issues, I was able to get it working with [this controller](https://gist.github.com/bkeroackdsc/c282b8473fd35127eee2).

I made the following changes:

- Commented out the secret volume, removed the volume mount from the kube2sky container.
- Removed the ``kube_cfg`` command line option from kube2sky (was referencing secret volume)
- Changed ``kube_master_url`` command line option to the insecure HTTP port (8080).
